_drafts/2024-09-01-aws-terraform-02.md

```diff
@@ -0,0 +1,267 @@
+
+```terraform
+# /infra/unicorn-api.tf
+
+resource "aws_iam_instance_profile" "unicorn_api_iam_profile" {
+  name = "unicorn-api-iam-profile"
+  role = aws_iam_role.ec2_readonly_role.name
+}
+```
+
+
+Now add the [iam instance profile](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/iam_instance_profile) to the aws instance.
+
+```terraform
+# /infra/unicorn-api.tf
+
+resource "aws_instance" "unicorn_api" {
+
+  # ...
+
+  iam_instance_profile = aws_iam_instance_profile.unicorn_api_iam_profile.name
+ 
+  # ...
+}
+```
+
+Great! Now that we've defined the IAM role and applied it to our EC2 resource, we can define a script for our instance to run on start up which will pull the ECR image and run it in docker.
+
+Create a new file in the `infra` directory, `init-unicorn-api.sh`
+
+```bash
+#!/bin/bash
+
+aws ecr get-login-password --region [region] | docker login --username AWS --password-stdin [ecr repository url]
+docker run -p 80:8080 [ecr repository url]:latest
+```
+
+Now add the following permissions for the script:
+
+`chmod 755 init-unicorn-api.sh`
+
+This give the file owner read, write, and execute permissions and everyone else read and execute permissions.
+
+Now add the script to the EC2 instance user data property:
+
+```terraform
+#! /infra/unicorn-api.tf
+
+resource "aws_iam_instance_profile" "unicorn_api_iam_profile" {
+  # ...
+  
+  user_data = file("${path.module}/init-unicorn-api.sh")
+
+  # ...
+}
+```
+
+With all that setup, we should now have an EC2 instance running a simple web server listening on port 80. There's just one final step to allow this service to be reachable from any device on the internet.
+
+## Step 4 - Allow any device on the internet to connect to our web server
+
+So far we've created a web service Docker image, pushed the image to ECR, and pulled & run the image on our EC2 instance.  
+
+Now we need to revisit our security group rules and update our ingress rule to allow incoming connections from any IP over port 80. Remove your existing ingress rule and replace it with the following.
+
+```terraform
+# /infra/unicorn-api.tf
+
+resource "aws_vpc_security_group_ingress_rule" "allow_incoming_http" {
+  security_group_id = aws_security_group.unicorn_api_security_group.id
+  description       = "allow ingress via port 80"
+  cidr_ipv4         = "0.0.0.0/0"
+  ip_protocol       = "tcp"
+  from_port         = 80
+  to_port           = 80
+}
+
+```
+
+With all of that we can run `terraform apply` to apply our changes.  
+
+Copy the EC2 public domain name output into a browser (you may need to prepend `http://` to the address as your browser might assume to use https) and see your awesome website!
+
+![a really really really cool website](/assets/getting-started-with-aws-tf-02/hello-world.png)
+
+In our next post, we'll do some cleanup of our current infrastructure and define our own VPC and subnet instead of using the default resources.
```

Comment:
Not sure if this is enough for the next post... Maybe need to chat about what the next steps ought to be...